### PR TITLE
Questions PR - will not merge this one

### PR DESCRIPTION
**Pull request by daniela-opti**
### Description of Changes

Exploration - make requests to preference center api and expose methods for getting and setting preferences. A lot of these changes were made to help me understand how the sdk works, and I expect the final version to be very different. The purpose of this PR is to ask questions on some specific examples

### Breaking Changes

...

### Prepare:

- [ ] Detail any breaking changes. Breaking changes require a new major version number
- [ ] Check `pod lib lint` passes
- [ ] Update any relevant sections of the repository wiki pages on a branch

### Bump versions in:

- [ ] `OptimoveCore.podspec`
- [ ] `OptimoveNotificationServiceExtension.podspec`
- [ ] `OptimoveSDK.podspec`

- [ ] `OptimoveCore/Sources/Classes/Constants/SDKVersion.swift`

- [ ] `README.md`
- [ ] `CHANGELOG.md`

- [ ] Update major version numbers in wiki (basic integration + push guides)

### Integration tests

*T&T Only*

- [ ] Init SDK with only T&T credentials
- [ ] Associate customer
- [ ] Associate email
- [ ] Track events

*Mobile Only*

- [ ] Init SDK with all credentials
- [ ] Track events
- [ ] Associate customer (verify both backends)
- [ ] Register for push
- [ ] Opt-in for In-App
- [ ] Send test push
- [ ] Send test In-App
- [ ] Receive / trigger deep link handler (In-App/Push)
- [ ] Receive / trigger the content extension, render image and action buttons for push
- [ ] Verify push opened handler

*Deferred Deep Links*

- [ ] With app installed, trigger deep link handler
- [ ] With app uninstalled, follow deep link, install test bundle, verify deep link read from Clipboard, trigger deep link handler

*Combined*

- [ ] Track event for T&T, verify push received
- [ ] Trigger scheduled campaign, verify push received
- [ ] Trigger scheduled campaign, verify In-App received

### Release:

- [ ] Squash and merge to master
- [ ] Delete branch once merged
- [ ] Create tag from master matching chosen version
- [ ] Run `pod trunk push` to publish to CocoaPods

### Post Release:

- [ ] Push wiki pages to master

### CODE REVIEW

File: OptimoveCore/Sources/Classes/NetworkClient/NetworkRequest.swift [hunk 1]
```diff
@@ -87,10 +87,14 @@ public extension HTTPHeader {
     enum Fields: String {
         case contentType = "Content-Type"
         case userAgent = "User-Agent"
+        case tenantId = "X-Tenant-Id"
+        case accept = "accept"
```
Comment (eligutovsky) [inline OptimoveCore/Sources/Classes/NetworkClient/NetworkRequest.swift]: In Swift, you don't need to provide the same string value explicitly

File: OptimoveCore/Sources/Classes/NetworkClient/NetworkRequest.swift [hunk 2]
```diff
@@ -87,10 +87,14 @@ public extension HTTPHeader {
...
     }
 
     enum Values: String {
         case json = "application/json"
+        case tenantId = ""
```
Comment (eligutovsky) [inline OptimoveCore/Sources/Classes/NetworkClient/NetworkRequest.swift]: This case can be changed in a way to accept an arg with string

File: OptimoveSDK/Sources/Classes/Optimobile/PreferenceCenter.swift [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    private var networkClient: NetworkClient
+    private var storage: KeyValueStorage
+    
+    init(storage: KeyValueStorage, networkClient: NetworkClient) {
+        self.networkClient = networkClient
+        self.storage = storage
+    }
+    
+    private func getTenantId() throws -> String {
+        if let tenantIdValue = storage.value(for: .tenantID) as? Int {
```
Comment (eligutovsky) [inline OptimoveSDK/Sources/Classes/Optimobile/PreferenceCenter.swift]:
```suggestion
        if let tenantId: Int = storage.value(for: .tenantID) {
```

Comment (eligutovsky) [reply OptimoveSDK/Sources/Classes/Optimobile/PreferenceCenter.swift]: or check if the storage has `getTenantId`

File: OptimoveSDK/Sources/Classes/Optimobile/PreferenceCenter.swift [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                ],
+                queryItems:[
+                    URLQueryItem(name: "customerId", value: customerId),
+                    URLQueryItem(name: "brandGroupId", value: brandGroupId)
+                ]
+            )
+            
+            
+            let data = try await networkClient.performAsync(request)
+            let preferences = try JSONDecoder().decode(Preferences.self, from: data!)
```
Comment (eligutovsky) [inline OptimoveSDK/Sources/Classes/Optimobile/PreferenceCenter.swift]: Explicitly unwrapping `data!` is a dangerous operation. Let check a data absence on a line before with:
```swift
let data: Data = try cast(await networkClient.performAsync(request))
```

File: OptimoveSDK/Sources/Classes/Optimove.swift [hunk 5]
```diff
@@ -42,28 +42,28 @@ typealias Logger = OptimoveCore.Logger
...
         }
-
+        
         if config.isOptimobileConfigured() {
             shared.container.resolve { serviceLocator in
                 do {
                     let visitorId = try serviceLocator.storage().getInitialVisitorId()
                     let userId = try? serviceLocator.storage().getCustomerID()
-
+
```
Comment (eligutovsky) [inline OptimoveSDK/Sources/Classes/Optimove.swift]: hmm, all these empty lines can be disabled from xcode -> pref -> text editing. there should be some flag about it

File: OptimoveSDK/Sources/Classes/Optimove.swift [hunk 6]
```diff
@@ -111,7 +111,54 @@ public extension Optimove {
     @objc static func reportEvent(name: String, parameters: [String: Any] = [:]) {
         shared.reportEvent(name: name, parameters: parameters)
     }
+    
+    
+ @available(iOS 13.0, *)
+    @objc func getPreferencesAsync() {
```
Comment (eligutovsky) [inline OptimoveSDK/Sources/Classes/Optimove.swift]: Consider removing `async` from the method name. This method does not intend to expose the inner implementation.

**Review by eligutovsky — COMMENTED** [review]
Hey @dani-optimove,
I have put some comments on code. I hope it will help)

File: OptimoveCore/Sources/Classes/NetworkClient/NetworkClient.swift [hunk 7]
```diff
@@ -4,6 +4,22 @@ import Foundation
...
+                case .success(let response):
+                    continuation.resume(returning: response.body)
+                case .failure(let error):
+                    continuation.resume(throwing: error)
+                }
+            }
+        }
+    }
+}
+
```
Comment (daniela-opti) [inline OptimoveCore/Sources/Classes/NetworkClient/NetworkClient.swift:22]: I'm not entirely sure that this was the way to go, I was mostly trying to stop the errors from coming 😅 
What do you think about this @eligutovsky

File: OptimoveCore/Sources/Classes/NetworkClient/NetworkRequest.swift [hunk 8]
```diff
@@ -87,17 +87,32 @@ public extension HTTPHeader {
...
+        var value: String {
+            switch self {
+            case .json:
+                "application/json"
+            case .tenantId(let id):
+                id
+            case .textplain:
+                "text/plain"
+            }
+        }
```
Comment (daniela-opti) [inline OptimoveCore/Sources/Classes/NetworkClient/NetworkRequest.swift:108]: Not sure if this was the correct way to implement the argument passing, let me know

File: OptimoveSDK/Sources/Classes/Optimove.swift [hunk 9]
```diff
@@ -88,57 +89,117 @@ typealias Logger = OptimoveCore.Logger
...
             let tenantEvent = TenantEvent(name: event.name, context: event.parameters)
             serviceLocator.pipeline().deliver(.report(events: [tenantEvent]))
         }
     }
-
-    /// Report the event to Optimove SDK.
-    ///
-    /// - Parameters:
-    ///   - event: Instance of OptimoveEvent type.
+
```
Comment (daniela-opti) [inline OptimoveSDK/Sources/Classes/Optimove.swift]: currently working on figuring out how to fix the indentation

**Pull request closed without merging**